### ISSUE TITLE
[RFC] APEX style fused optimizers in PyTorch

### ISSUE THREAD

**Issue by crcrpar** [pytorch/pytorch#68041]
Having APEX style fused optimizers in pytorch by implementing, e.g., `_fused_sgd(Tensor[] params, Tensor[] grads, Tensor[] momentums, ...)` and expose this function via `torch.optim.fused.FusedSGD`.

## Motivation
We've developed `foreach` and `_multi_tensor` optimizers which use `foreach` functions to reduce the overhead of 
calling a bunch of small CUDA kernels.
As `foreach` handles one or more lists of tensors and optional scalar or a list of scalars, for dense contiguous tensors,
they are more efficient than the native optimizers (`torch.optim` optimizers) by the factor of the number of parameters in a model **without** much loss of flexibility and hackability (see https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/_functional.py for the implementation).

If we're sure that tensors involved in parameter update suffice the following conditions, [APEX style fused optimizers](https://github.com/NVIDIA/apex/tree/master/apex/optimizers)
should be more efficient than the `_multi_tensor` optimizers because each fused optimizer kernel takes list of tensors and do all the required math in it.

conditions
- tensors are dense and contiguous
- tensors are on the same device
- tensors have the same dtype

I'd love to know what you think.

rel: https://github.com/pytorch/pytorch/issues/38655

cc @ptrblck @mcarilli @ngimel

**Comment by ngimel** [comment]
cc @jbschlosser @albanD

**Comment by ngimel** [comment]
Also see #71274

**Comment by crcrpar** [comment]
found an old issue: https://github.com/pytorch/pytorch/issues/3929
